File: Data Structures & Algorithms/two-integer-sum/submission-2.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # 3
        # [0, 1, 2]
        # {0 : 0, 1 : 1, }
        # 3 - 2 = 1 -> dict[1]=> 1 => return [1, 2]
        mapping = dict()
        for i, x in enumerate(nums):
            if mapping.get(target - x, -1) != -1:
                return [mapping[target - x], i]
            mapping[x] = i
        
        return []

File: Data Structures & Algorithms/binary-search/submission-2.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        return binary_search(nums, 0, nums.size() - 1, target);
    }
    int binary_search(vector<int>& nums, int l, int r, int target) {
        if (l > r) return -1;
        int m = l + (r - l) / 2;
        if (nums[m] == target) return m;
        if (nums[m] > target) return binary_search(nums, l, m - 1, target);
        return binary_search(nums, m + 1, r, target);
    }
};
